### ISSUE TITLE
Solving errors when reading DataMatrix Codes

### ISSUE THREAD

**Issue by fabulous-fabio** [zxing-cpp/zxing-cpp#782]
Hello,
I'm currently working on a Project where I have to decode many small DataMatrixCodes.
I'm using the python wrapper with a self built wheel using c++20.
The Images have a relatively low resolution, but I can still read about half the codes.
Unfortunately I encountered some problems:
When scanning the rest of the codes, I eather encounter a ChecksumError or an FormatError.

I already tried to rescale the images but I still encounter the errors.
When I read the codes with an App that is also using this library it reads the codes well. Does anybody know how I can achieve this?
This Codes raises the following Error: ChecksumError @ DMDecoder.cpp:425

![img_17](https://github.com/zxing-cpp/zxing-cpp/assets/120569956/bafe2493-2887-493f-aa07-f1cf56e3551a)
![img_19](https://github.com/zxing-cpp/zxing-cpp/assets/120569956/36a85e87-8790-48dc-95ed-fe75dc2bf72f)

And this Codes raises the following Error: FormatError (Invalid matrix dimension) @ DMDecoder.cpp:399
![img_13](https://github.com/zxing-cpp/zxing-cpp/assets/120569956/fc388a3e-18c8-4f53-9ed0-f85522b8056a)
![img_12](https://github.com/zxing-cpp/zxing-cpp/assets/120569956/f8aeaf9c-5669-4a71-bbf0-46bd026c6142)

Maybe someone can help me figure out how I have to manipulate the image to be able to read the codes.

Thank you a lot!
Kind Regards

**Comment by axxel** [comment]
Your 4 images don't work because of a combination of 3 problems:
1) low resolution
2) background noise
3) over saturation (black dots are too thick)

They are very similar in nature. Do all your problematic images look like this?

I had some success in detecting them by doing: a) scale them up to 200%, then b) apply an optional dilate operation and c) perform a thresholding such that you get roughly the same amount of black and white pixels within the symbol. If your input is really so similar, you could do the latter by hand once and then use the same thresholding value for all your input.

Here are my results that work:
![#782-1](https://github.com/zxing-cpp/zxing-cpp/assets/1038346/910f77a9-5f6f-489f-a304-79fe1e539c3e)
![#782-3](https://github.com/zxing-cpp/zxing-cpp/assets/1038346/d5e63308-1936-4718-8d34-f8e87505f26e)
![#782-4](https://github.com/zxing-cpp/zxing-cpp/assets/1038346/e0d632fd-65a0-4542-9c26-34becbe92c3a)

**Comment by axxel** [comment]
I forgot to ask: which app based on this library were you talking about?

**Comment by fabulous-fabio** [comment]
Thanks for your answer!
What Kernel did you use for dilation?

I'll try it out and will let you know how it works.

The images are fairly similar because they are all from a bigger image and thus all have the same quality issues.

The app I use is calles Code Scan

Kind Regards

**Comment by axxel** [comment]
> What Kernel did you use for dilation?

I simply used GIMP. I believe they use a 3x3 box kernel.

> The app I use is calles Code Scan

I can't find it, can you provide a URL?

**Comment by fabulous-fabio** [comment]
Hey, glad to inform you that the provided Images work fairly well now.

The app mentioned is the following: https://apps.apple.com/de/app/code-scan-liest-jeden-qrcode/id1554812545

I still face some issues I hope you can help me with. The following codes fail to decode and I hope you can assist me with your valuable expertise:

![img_12](https://github.com/zxing-cpp/zxing-cpp/assets/120569956/fad7a321-1f05-4bc1-81b7-cd71c556c300)
![img_9](https://github.com/zxing-cpp/zxing-cpp/assets/120569956/5c82d021-4c30-4004-b77e-9d6921376446)
![img_8](https://github.com/zxing-cpp/zxing-cpp/assets/120569956/8b9a7eef-f30c-4e93-94e9-732f0d295f95)
![img_7](https://github.com/zxing-cpp/zxing-cpp/assets/120569956/4bd372b5-c143-4736-9616-0bada0222e0c)
![img_6](https://github.com/zxing-cpp/zxing-cpp/assets/120569956/0b6558b0-b92f-4903-ae85-6cc35ee4558b)
![img_5](https://github.com/zxing-cpp/zxing-cpp/assets/120569956/0dbbd2c3-0abb-40e6-bca2-6f9e155b9e75)
![img_4](https://github.com/zxing-cpp/zxing-cpp/assets/120569956/84076b8a-0350-4efe-9f1f-d0e558c021e1)
![img_2](https://github.com/zxing-cpp/zxing-cpp/assets/120569956/d3afd733-d500-4d89-b690-73a9d34787e6)
![img_0](https://github.com/zxing-cpp/zxing-cpp/assets/120569956/eced5c3a-63e0-423f-a57c-ec6bcb8cdb1d)

Thanks for your help and the development of this very helpful library!

Kind regards

**Comment by axxel** [comment]
So you implemented the 3 steps I mentioned above in your application and it helps but the provided images still have issues?

With regards to why the mentioned iOS app works better: Let's ask the authors, maybe they can share their secret source? ;) @parallaxe @markusfisch

**Comment by fabulous-fabio** [comment]
> So you implemented the 3 steps I mentioned above in your application and it helps but the provided images still have issues?
> 
> 
> 
> With regards to why the mentioned iOS app works better: Let's ask the authors, maybe they can share their secret source? ;) @parallaxe @markusfisch 

Exactly. A lot more Images work now, but somehow these images still don't.

**Comment by fabulous-fabio** [comment]
Does anybody have an Idea how to solve them?
Kind regards

**Comment by Eitol** [comment]
The problem is occurring because neither of the 2 functions that search for the L pattern support this pattern being discontinuous.
That's why it throws the error "FormatError (Invalid matrix dimension)".

The function must be modified to detect the L pattern with discontinuous lines.

static DetectorResult Scan(EdgeTracer& startTracer, std::array<DMRegressionLine, 4>& lines)

I'm going to try to do it (But I'm not an expert on this topic)

![image](https://github.com/user-attachments/assets/4f343692-3fa8-4e1e-b97d-ba2d74f378d3)

**Comment by axxel** [comment]
The problem discussed at the end is related to #234.
